### PR TITLE
MAILBOX-294 Avoid updating flags when not needed (Cassandra)

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/FunctionRunnerWithRetry.java [hunk 1]
```diff
@@ -32,8 +32,97 @@
...
     @FunctionalInterface
-    public interface OptionalSupplier<T> {
-        Optional<T> getAsOptional();
+    public interface ExecutionResultSupplier<T> extends Supplier<ExecutionResult<T>> {}
+
+    public interface ExecutionResult<T> {
+        boolean failed();
+        boolean canceled();
+        boolean succeeded();
+        Optional<T> value();
```
Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/FunctionRunnerWithRetry.java]: asOptional

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/FunctionRunnerWithRetry.java [hunk 2]
```diff
@@ -32,8 +32,97 @@
     public static class RelayingException extends RuntimeException {}
 
     @FunctionalInterface
-    public interface OptionalSupplier<T> {
-        Optional<T> getAsOptional();
+    public interface ExecutionResultSupplier<T> extends Supplier<ExecutionResult<T>> {}
+
+    public interface ExecutionResult<T> {
+        boolean failed();
```
Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/FunctionRunnerWithRetry.java]: I would have use a status enum

File: backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/CassandraCluster.java [hunk 3]
```diff
@@ -55,7 +55,9 @@ public static CassandraCluster create(CassandraModule module) throws RuntimeExce
     private CassandraCluster(CassandraModule module, EmbeddedCassandra embeddedCassandra) throws RuntimeException {
         this.module = module;
         try {
-            session = new FunctionRunnerWithRetry(MAX_RETRY).executeAndRetrieveObject(CassandraCluster.this::tryInitializeSession);
+            session = new FunctionRunnerWithRetry(MAX_RETRY)
+                .executeAndRetrieveObject(CassandraCluster.this::tryInitializeSession)
+                .get();
```
Comment (mbaechler) [inline backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/CassandraCluster.java]: get is evil, you should get or throw instead

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java [hunk 4]
```diff
@@ -248,30 +248,34 @@ public void delete(MessageId messageId) {
...
-    private Optional<Pair<Flags, ComposedMessageIdWithMetaData>> updateFlags(MailboxId mailboxId, MessageId messageId, Flags newState, MessageManager.FlagsUpdateMode updateMode) throws MailboxException {
+    private FunctionRunnerWithRetry.ExecutionResult<Pair<Flags, ComposedMessageIdWithMetaData>> updateFlags(MailboxId mailboxId, MessageId messageId, Flags newState, MessageManager.FlagsUpdateMode updateMode) throws MailboxException {
         CassandraId cassandraId = (CassandraId) mailboxId;
         ComposedMessageIdWithMetaData oldComposedId = imapUidDAO.retrieve((CassandraMessageId) messageId, Optional.of(cassandraId))
             .join()
             .findFirst()
             .orElseThrow(MailboxDeleteDuringUpdateException::new);
+        Flags newFlags = new FlagsUpdateCalculator(newState, updateMode).buildNewFlags(oldComposedId.getFlags());
+        if (oldComposedId.getFlags().equals(newFlags)) {
+            return FunctionRunnerWithRetry.canceled();
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java]: it should not be canceled, it's just an immediate success

**Comment by chibenwa** [comment]
Then cancelation feature should not be needed, and we might avoid troubbles in Concurrent MPT tests. Will PR again.

**Pull request closed without merging**